### PR TITLE
feat:  Add an error variant RetryLater

### PR DESCRIPTION
**Pull request by jun0315**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add an error variant RetryLater for recoverable failure. (The default variant External always abort the procedure.)

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
https://github.com/GreptimeTeam/greptimedb/issues/286

## TODO
In order to better manage the retry logic, we will increase the maximum number of retries and increase the waiting time exponentially

### CODE REVIEW

File: src/common/procedure/src/local/runner.rs [hunk 1]
```diff
@@ -213,6 +217,10 @@ impl Runner {
                     self.meta.id
                 );
 
+                if let Error::RetryLater { .. } = e {
```
Comment (evenyag) [inline src/common/procedure/src/local/runner.rs:220]: Print whether the error is RetryLater in the log above.
```
"Failed to execute procedure {}-{}, retry: {}"
....
e.is_retry_later()
```

File: src/storage/Cargo.toml [hunk 2]
```diff
@@ -14,6 +14,7 @@ arrow-array.workspace = true
 bytes = "1.1"
 common-base = { path = "../common/base" }
 common-error = { path = "../common/error" }
+common-procedure = { path = "../common/procedure" }
```
Comment (evenyag) [inline src/storage/Cargo.toml:17]: Do we need this import?

**Comment by jun0315** [comment]
@evenyag PTAL :D

**Comment by evenyag** [comment]
I added a new error `table_procedure::Error` in [#1040](https://github.com/GreptimeTeam/greptimedb/pull/1040/files#diff-f9d96740736fa846ca764669918fd9fb79e1cf74753fece8d77a9f0ef755b11c)

Could you also refactor the `From` implementation for it?

File: src/common/procedure/src/error.rs [hunk 3]
```diff
@@ -111,4 +118,50 @@ impl Error {
...
+
+    /// Creates a new [Error::RetryLater] error from source `err`.
+    pub fn retry_later<E: ErrorExt + Send + Sync + 'static>(err: E) -> Error {
+        Error::RetryLater {
+            source: BoxedError::new(err),
+        }
+    }
+
+    /// Determine whether it is a retry later type through [StatusCode]
+    pub fn is_retry_later(status_code: StatusCode) -> bool {
```
Comment (evenyag) [inline src/common/procedure/src/error.rs:130]: This reminds me that we could add a method like `is_retryable` to the StatusCode

**Comment by jun0315** [comment]
> I added a new error `table_procedure::Error` in [#1040](https://github.com/GreptimeTeam/greptimedb/pull/1040/files#diff-f9d96740736fa846ca764669918fd9fb79e1cf74753fece8d77a9f0ef755b11c)
> 
> Could you also refactor the `From` implementation for it?

Good idea! I have solved all the above code reviews.:D

**Review by evenyag — APPROVED** [review]
LGTM

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**